### PR TITLE
Reland: Plumb dart timeline events to mojo:tracing.

### PR DESCRIPTION
**Pull request by ppiet**
This updates and relands https://github.com/flutter/engine/commit/ef589f99bca883bf27a2596f75a531c7fd517774 which was reverted due to a build-breaking conflict with a recent mojo roll. Please find the original description below.

This patch:

 - makes dart traces (coming from the VM and the timeline API)
   available to mojo:tracing
 - adds --vm-complete-timeline parameter, which makes the Dart VM enable all
   categories of timeline traces from the beginning, allowing to do startup
   tracing

This is based on the equivalent support in dart content handler.

Fixes https://github.com/flutter/flutter/issues/3162.

### CODE REVIEW

**Comment by ppiet** [comment]
Note that one can't currently test if this works as mojo:tracing seems to be broken, see https://github.com/domokit/mojo/issues/775 . Please consider this on hold - I will confirm that it works and update this thread when mojo tracing is fixed and rolled into engine.

cc @tonygentilcore

**Comment by eseidelGoogle** [comment]
Since @abarth is out, I would like @johnmccutchan to explicitly sign off on this patch and merge it.

**Comment by johnmccutchan** [comment]
@ppiet what is the delta between this version and the one that broke the build?

**Comment by ppiet** [comment]
@johnmccutchan please find the diff attached. This corresponds to the changes to dart_tracing.{h,cc} made in mojo in  https://codereview.chromium.org/1918333003 and https://codereview.chromium.org/1928183002 .

[diff.txt](https://github.com/flutter/engine/files/261891/diff.txt)

File: sky/engine/core/script/dart_init.cc [hunk 1]
```diff
@@ -114,6 +114,10 @@ static const char* kDartTraceStartupArgs[]{
     "--timeline_recorder=endless",
 };
 
+static const char* kVmCompleteTimelineArgs[]{
```
Comment (johnmccutchan) [inline sky/engine/core/script/dart_init.cc]: Why is this necessary? The kDartTraceStartupArgs should give you what you need.

File: sky/shell/platform/mojo/dart_tracing.cc [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+#include "mojo/public/cpp/bindings/interface_request.h"
+#include "mojo/services/tracing/interfaces/trace_provider_registry.mojom.h"
+
+namespace dart {
+
+void DartTimelineController::Enable(const mojo::String& categories) {
+  if (categories == mojo::String("Dart")) {
+    Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_DART);
+  } else {
+    // TODO(johnmccutchan): Respect |categories|.
```
Comment (johnmccutchan) [inline sky/shell/platform/mojo/dart_tracing.cc]: s/johnmccutchan/ppiet.

File: sky/shell/platform/mojo/dart_tracing.cc [hunk 3]
```diff
@@ -0,0 +1,172 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (johnmccutchan) [inline sky/shell/platform/mojo/dart_tracing.cc]: This file is basically the same as the one in Mojo, right? Is there a way to share the code?

File: sky/shell/platform/mojo/main_mojo.cc [hunk 4]
```diff
@@ -16,12 +16,14 @@
...
 #include "sky/shell/shell.h"
 #include "sky/shell/platform/mojo/content_handler_impl.h"
+#include "sky/shell/platform/mojo/dart_tracing.h"
 
 namespace sky {
 namespace shell {
 namespace {
 
 const char kEnableCheckedMode[] = "--enable-checked-mode";
+const char kVmCompleteTimeline[] = "--vm-complete-timeline";
```
Comment (johnmccutchan) [inline sky/shell/platform/mojo/main_mojo.cc]: Flutter already has --trace-startup which will give you the complete timeline (minus events you don't care about).

File: sky/shell/platform/mojo/dart_tracing.cc [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+// 2. Expects to receive one or more complete JSON maps per call.
+// Therefore, we parse the recorded data, split it up and send it to trace
+// recorder in chunks.
+void DartTraceProvider::SplitAndRecord(char* data, size_t length) {
+  const size_t kMinChunkLength = 1024 * 1024;  // 1MB.
+
+  // The data from the VM is null-terminated content of a JSON array without its
+  // enclosing square brackets. Remove the trailing "\0" and add the enclosing
+  // brackets to make JSON parser happy.
+  std::string json_data = "[" + std::string(data, length -1) + "]";
```
Comment (abarth) [inline sky/shell/platform/mojo/dart_tracing.cc]: This is very inefficient.

File: sky/shell/platform/mojo/dart_tracing.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#define SKY_SHELL_PLATFORM_MOJO_DART_TRACING_H_
+
+#include "base/trace_event/trace_event.h"
+#include "mojo/common/tracing_impl.h"
+#include "mojo/public/cpp/bindings/interface_request.h"
+#include "mojo/services/tracing/interfaces/tracing.mojom.h"
+
+namespace dart {
+
+class DartTimelineController {
```
Comment (abarth) [inline sky/shell/platform/mojo/dart_tracing.h]: One class per file please.

File: sky/shell/platform/mojo/dart_tracing.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+#ifndef SKY_SHELL_PLATFORM_MOJO_DART_TRACING_H_
+#define SKY_SHELL_PLATFORM_MOJO_DART_TRACING_H_
+
+#include "base/trace_event/trace_event.h"
+#include "mojo/common/tracing_impl.h"
+#include "mojo/public/cpp/bindings/interface_request.h"
+#include "mojo/services/tracing/interfaces/tracing.mojom.h"
+
+namespace dart {
```
Comment (abarth) [inline sky/shell/platform/mojo/dart_tracing.h]: This namespace isn't correct.  Everything else in this directory is in the `sky::shell` namespace.

File: sky/engine/core/script/dart_init.cc (hunk 1 above)
Comment (ppiet) [inline sky/engine/core/script/dart_init.cc]: Nice, done. ("--trace-startup" used not to set timeline recorder to endless back at the time of the first patchset)

File: sky/shell/platform/mojo/dart_tracing.cc (hunk 3 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.cc]: The purpose is the same. We can put this under mojo/common - but since this depends on dart/, we would need to roll dart into engine in lockstep with mojo. Or maybe dart_tools_api.h is stable enough that this will not be a problem in practice?

File: sky/shell/platform/mojo/dart_tracing.cc (hunk 2 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.cc]: Done.

File: sky/shell/platform/mojo/dart_tracing.h (hunk 7 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.h]: Done.

File: sky/shell/platform/mojo/dart_tracing.h (hunk 6 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.h]: Done. Also made all these static methods just global functions in anonymous namespace and dropped DartTimelineController.

File: sky/shell/platform/mojo/main_mojo.cc (hunk 4 above)
Comment (ppiet) [inline sky/shell/platform/mojo/main_mojo.cc]: Yay, done.

File: sky/shell/platform/mojo/dart_tracing.cc (hunk 5 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.cc]: Switched to std::stringstream.

**Comment by ppiet** [comment]
Thanks, @johnmccutchan and @abarth ! I addressed your comments, but haven't tested the newest patchset yet - waiting for the mojo fix for tracing to roll into engine and elsewhere. You might want to hold off until then, I will update this thread when this is verified.

**Comment by abarth** [comment]
I'm working on a mojo roll in https://github.com/flutter/engine/pull/2717.  Just waiting for some iOS issues to stablize before landing.

**Comment by Hixie** [comment]
@ppiet did you have a chance to look at this since the mojo roll?

**Comment by ppiet** [comment]
@Hixie no. Mojo with the tracing fix rolled into Flutter (thanks!), but not yet into the consumer of this feature.

**Comment by Hixie** [comment]
@ppiet Ok. I'm going to close this PR for now so that it isn't on our review-needed radar, please don't hesitate to reopen it when you're ready!
